blog/2025-06-10-esm-scripts-are-here.md

```diff
@@ -0,0 +1,94 @@
+---
+authors: mark
+comments: false
+date: 2025-06-09 10:00:00+01:00
+layout: post
+slug: esm-scripts-playcanvas-2025
+title: "ESM Scripts"
+categories:
+  - Announcement
+  - Engine
+  - Scripting
+tags:
+  - esm
+  - modules
+  - editor
+  - developer
+---
+
+## Modern JavaScript in PlayCanvas
+
+**We’re excited to announce a major step forward for the PlayCanvas editor: **ESM Scripts are now supported** — and officially the recommended way to write scripts in your PlayCanvas projects.**
+
+If you’ve ever struggled with managing classic scripts, wondered why your auto-complete didn’t work, or wished you could reuse code across projects more easily — this is for you.
+
+ESM (ECMAScript Modules) brings modern JavaScript development to the heart of PlayCanvas. Think: cleaner code, better tooling, instant editor feedback, and full compatibility with the broader JS ecosystem. It’s faster to get started, easier to scale your project, and way more fun to work with.
+
+No more hidden globals. No more messy script loading order. Just well-structured, maintainable, and modular code — exactly how modern web development should be.
+
+<video src="/img/pc-esm-scripts.mp4" autoPlay type="video/mp4" width="100%" />
+
+---
+
+## Why use ESM Scripts?
+
+ESM Scripts offer major improvements across the board:
+
+💡 **Modern JavaScript** — use `import`, `export` class based module syntax  
+🧠 **Smarter Editor** — better auto-complete and inline docs  
+🧱 **Modular Codebase** — structure large projects cleanly with reusable modules  
+⚙️ **No Global Scope Issues** — each script has its own context  
+📦 **Import Maps** — define aliases and pull in libraries from CDNs  
+🚀 **Optimized for Production** — static imports enable tree-shaking and bundling
```

Comment:
Bit concerned about this. We don't do tree-shaking, at least not from Editor projects. Feel free to outline your vision for how tree shaking will work in the future though?